### PR TITLE
Clearer error message for plt.axis()

### PR DESCRIPTION
**Pull request by dstansby**
Fixes https://github.com/matplotlib/matplotlib/issues/13668

### CODE REVIEW

File: lib/matplotlib/axes/_base.py [hunk 1]
```diff
@@ -1660,11 +1660,15 @@ def axis(self, *args, emit=True, **kwargs):
...
+                limits = args[0]
                 try:
-                    xmin, xmax, ymin, ymax = v
+                    xmin, xmax, ymin, ymax = limits
                 except ValueError:
                     raise ValueError('args must contain [xmin xmax ymin ymax]')
+                except TypeError:
+                    raise TypeError('the first argument to axis() must be an '
+                                    'interable of the form '
+                                    '[xmin, xmax, ymin, ymax]')
```
Comment (jklymak) [inline lib/matplotlib/axes/_base.py]: Not entirely following why one error would be triggered versus the other, but can the same string be used for each?

Comment (dstansby) [reply lib/matplotlib/axes/_base.py]: They're different errors; a `TypeError` is thrown if iterable unpacking fails, whereas I'm not actually sure what exact case the previous one is catching.

Comment (jklymak) [reply lib/matplotlib/axes/_base.py]: OK, but it seems the error string should say the same thing in both cases?

Comment (anntzer) [reply lib/matplotlib/axes/_base.py]: valueerror is raised if then length is not correct.

Comment (dstansby) [reply lib/matplotlib/axes/_base.py]: Yeah, I think it's reasonable for the error message to be the same

File: lib/matplotlib/axes/_base.py [hunk 2]
```diff
@@ -1660,11 +1660,13 @@ def axis(self, *args, emit=True, **kwargs):
...
                         "argument to axis() is deprecated and will raise a "
                         "TypeError %(removal)s.")
-                v = args[0]
+                limits = args[0]
                 try:
-                    xmin, xmax, ymin, ymax = v
-                except ValueError:
-                    raise ValueError('args must contain [xmin xmax ymin ymax]')
+                    xmin, xmax, ymin, ymax = limits
+                except TypeError, ValueError:
```
Comment (anntzer) [inline lib/matplotlib/axes/_base.py]: need parentheses

File: lib/matplotlib/axes/_base.py [hunk 3]
```diff
@@ -1660,11 +1660,13 @@ def axis(self, *args, emit=True, **kwargs):
...
+                    raise TypeError('the first argument to axis() must be an '
+                                    'interable of the form '
```
Comment (timhoffm) [inline lib/matplotlib/axes/_base.py]:
```suggestion
                                    'iterable of the form '
```

**Comment by QuLogic** [comment]
Might have to rebase for Pandas fix?

**Pull request merged**